### PR TITLE
#175 - Finished testing InMemoryInvoices.

### PR DESCRIPTION
**Pull request by criske**
PR for #175 

Left puzzle to refactor and test ContractInvoices.

### CODE REVIEW

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java [hunk 1]
```diff
@@ -134,33 +138,153 @@ public void paysInvoice(){
...
+            Contract.Roles.ARCH
+        );
+        storage.invoices().add(this.mockAssignedTask(
+            "123", storage, project, contributor, Contract.Roles.DEV),
+            Duration.ofHours(1));
+        storage.invoices().add(this.mockAssignedTask(
+            "124", storage, project, contributor, Contract.Roles.ARCH),
+            Duration.ofHours(2));
 
+        storage.invoices().ofContract(contractIdTwo).isPaid(1);
```
Comment (amihaiemil) [inline self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java:238]: @criske What is the ``1`` argument in ``.isPaid(1)``? Is it the InvoiceId?

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java [hunk 2]
```diff
@@ -134,33 +138,153 @@ public void paysInvoice(){
...
+            contributor.username(),
+            contributor.provider(),
+            Contract.Roles.ARCH
+        );
+        storage.invoices().add(this.mockAssignedTask(
+            "123", storage, project, contributor, Contract.Roles.DEV),
+            Duration.ofHours(1));
+        storage.invoices().add(this.mockAssignedTask(
+            "124", storage, project, contributor, Contract.Roles.ARCH),
+            Duration.ofHours(2));
```
Comment (amihaiemil) [inline self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java:284]: @criske Shouldn't this test fail here or when adding the first task above? Why does it fail only when trying to read the Task from teh contract (the last line  bellow)?

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java [hunk 3]
```diff
@@ -266,12 +273,6 @@ public boolean isPaid(final int id) {
                 throw new IllegalStateException("Invoice with id"
                     + id + " is not part of the contract.");
             }
-            return storage.invoices().tasks(id);
-        }
-
-        @Override
-        public Iterator<Invoice> iterator() {
-            return contractInvoices.get().iterator();
         }
```
Comment (amihaiemil) [inline self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java:276]: @criske Why remove the iterator?

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java [hunk 4]
```diff
@@ -193,11 +190,6 @@ public void pay(final int id) {
          */
         private final Supplier<Stream<Invoice>> contractInvoices;
 
-        /**
-         * Current active invoice.
-         */
-        private Integer activeId;
```
Comment (amihaiemil) [inline self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java:199]: @criske Why remove this?

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java (hunk 1 above)
Comment (criske) [reply self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java:238]: Yes

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java (hunk 2 above)
Comment (criske) [reply self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoicesTestCase.java:284]: Will not fail because we adding two tasks each linked with a different contract. (one contract has DEV role other ARCH, diff keys basically).

The last line will fail becase we're trying to get tasks of an invoice that is not part of the contract.
Invoice with id 1 is part of the contractOne not contractTwo.

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java (hunk 3 above)
Comment (criske) [reply self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java:276]: Is not removed, it will fail the build if so; was just displaced I guess.
https://github.com/self-xdsd/self-core/blob/b23b0cf6e44fa481ce403a926a44de8bb251cdac/self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java

File: self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java (hunk 4 above)
Comment (criske) [reply self-core-impl/src/test/java/com/selfxdsd/core/mock/InMemoryInvoices.java:199]: Was a bug with that aproach.
Will fail the test when we check if the invoice contract is paid. You have to check the actual storage if payment is done.
Here is the corner case :
`Invoices invoicesOfContract = contract.invoices() .ofContract(contractId);`
then we add contract task through overall invoices not through **invoicesOfContract**:
`storage.invoices().add(taskOfContract, duration)`
then we "pay"
`storage.invoices().pay(contractInvoiceId)`
Now if you call  `invoicesOfContract.isPaid(contractInvoiceId) ` the old code will return **false** instead of **true** (stale state of **activeId**).

**Comment by amihaiemil** [comment]
@criske thanks for the answers!

**Comment by amihaiemil** [comment]
@rultor merge
